### DISCUSSION TITLE
переход на platformio

### DISCUSSION THREAD

**Discussion by hhrhhr** [vvip-68/LedPanelWiFi#101] [Идеи и предложения]
Я в теме гирлянд/ламп/матриц 2-3 недели, наверное, всего, но вопросов типа "скомпилировал, прошил - нет веб интерфейса" уже встретил больше половины от общего числа. Может пора уже перейти окончательно и бесповоротно на использование [PlatformIO (Core)](https://docs.platformio.org/en/latest/core/index.html)?
IDE как таковое подавляющему большинству не нужно, да и ардуинские поделки называть IDE как-то совестно. Конечному пользователю достаточно будет отредактировать под себя 2 файла (a_def_*) и по очереди выполнить 3 команды:
```
pio run --target erase
pio run --target upload
pio run --target uploadfs
```
Всё, можно отключать и пользоваться.

PlatformIO Core (CLI) места практически не занимает, фреймворки и библиотеки добывает автоматически, с проектом достаточно держать "нештатные" ESP_Async_E1.31 (кем-то форкнутый без указания онлайн местонахождения) и ESPAsyncWebServer (и то dvarrel/ESPAsyncWebSrv@1.2.9 с минимальными правками работает на D1 Mini), остальное подтягивается из официального репозитория или по прямым ссылкам с гитхаба. На худой конец можно все проверенные библиотеки сложить в кучу локально, предварительно выкинув из них тесты/примеры и тд.

Если прям IDE будет необходима, то pio кроме (ненужно) VSCode умеет генерировать проектные файлы [ещё для десятка сред/редакторов](https://docs.platformio.org/en/latest/integration/ide/index.html#desktop-ides). Codeblocks или QTCreator на голову выше Arduino IDE, кроме, разве что, отсутствия встроенного монитора порта.

Сама же подготовка для первой компиляции ещё более простая, достаточно скачать один статически собранный бинарник [uv от Astral](https://github.com/astral-sh/uv/releases), через него установить Python и PlatformIO. Если питон уже есть, то uv ставится через pip/pipx, в общем вариантов много.

Со своей стороны могу предоставить конфигурацию для ESP8266, которая собирается и работает почти для всех вариантов, кроме MP3-плеера и TM1637 (по причине отсутствия). Полагаю, для ESP32 будет несложно подставить нужные библиотеки и ключи компиляции.

Плюсы: uv просто реактивен в плане установки питоновских пакетов-модулей; pio такой же быстрый при разрешении зависимостей, самой компиляции и перекомпиляции; настройка требует лишь ввода нескольких команд; к этой связке идеально подходит fastled-wasm для "отладки"; не нужно будет гадать, те ли версии библиотек/фреймворков поставил пользователь; можно включить полный вывод при компиляции и увидеть вообще все флаги и определения для каждого объекта.

Минусы: многопоточная компиляция и, как следствие, сложность отлова ошибок — компиляторы ссылаются не на исходники, а на огромный сборный .cpp-файл, в который попадают все файлы проекта.

Вот как-то так.

**Comment by vvip-68** [comment]
1. Кто мешает пользоваться PlatformIO прямо сейчас? Мне не нравится эта система / IDE - я не буду ей пользоваться.

2. Кто знает про эти три команды? Почему есть уверенность, что их будут просто выполнять? Уж с настройкой параметров (в т.ч и Arduino IDE) - куча текстов и инструкций написано, раздел WIKI есть, в основном файле скетча в самом начале - все разъяснено, какие версии нужны и т.д. Думаете их кто-то читает?  Также и с PlatformIO - никто ничего читать не будет.

Arduino IDE также автоматически подгружает все что нужно. И тут начинаются проблемы. Потому что то-что оно подгружает автоматически - для проекта по большей части не подходит от слова "совсем". Начиная с того, что часть библиотек слегка подправлены, потому что стандартные неправленные не работают как надо, заканчивая тем, что по умолчанию ставит FastLED 3.9.4 с которым проект либо не соберется вообще, либо на 8266 работать не будет.   

И ведь везде написано - не ставьте 3.9.4, написано какие нужны версии. Но опять же - кто это читает? И тут начинается поток вопросов - почему не компилится, почему не работает?.

По предложениям - выкладывайте свои наработки с описанием как ими пользоваться, Опишите все детально - разместим в Wiki проекта, добавим всё что необходимо.

**Reply by hhrhhr** [reply]
> Arduino IDE также автоматически подгружает все что нужно.

Нет, только ручками. Если не положить отсутствующую библиотеку в каталог library, будет ругань.
pio же, помимо загрузки и установки фреймворка для выбранной в конфиге платформы, проходит парсером по всем `#include` и пытается найти соответствие со своим репозиторием. Скажем, встретилось `#include "GyverButton.h"` - будет загружена и использована либа [gyverlibs/GyverButton](https://registry.platformio.org/libraries/gyverlibs/GyverButton). Или есть конкретно указать в конфиге `bblanchon/ArduinoJson@6.21.5`, то подключится она, а не распоследняя 7.2.1.

> часть библиотек слегка подправлены

И вот это очень плохо. Теряется связь с оригинальной библиотекой, обновить её без автора правок проблематично.
ESP_Async_E1.31, приложенная к проекту, всё ещё содержит ссылки на оригинальный репозиторий от forkineye, но "исправлена" на предмет не только получения E1.31 команд, но их отправки, как я понял. Для меня возможность управления другими гирляндами через саму гирлянду вообще просто классная фича, без задействования каких-либо промежуточных серверов/компов/ноутбуков.
Но вот pio, без пинка свыше, тащит оригинальную библиотеку и привет. А там поддержка передачи команд есть только в pull-реквесте, которому пару лет.

> по умолчанию ставит FastLED 3.9.4 с которым проект либо не соберется вообще

О чём и речь, одна строчка в _platfomio.ini_ вида `lib_deps = FastLED/FastLED@3.6.0` решает все проблемы и не даст пользователю устанавливать что попало.

> И ведь везде написано - не ставьте 3.9.4

Вы сами отвечали, типа "при всяких изменениях я что, должен проходить по всем инструкциям и картинкам и менять их на актуальные?". При использовании pio достаточно будет одной единственной инструкции с тремя командами для компиляции/прошивки (ну и ещё парой строк про установку uv/python/platfomio), и не надо будет распыляться на коментарии в коде, в вики и ещё на паре страниц, которые до сих противоречат друг другу ;)

**Reply by hhrhhr** [reply]
Основной посыл — не трогая существующий код просто адаптировать систему сборки, исключив пользовательские ошибки и невнимательность. В идеале же перейти от самостоятельной сборки к загрузке нужной готовой прошивки.
Да, вариантов прошивок многовато, но если чуть-чуть потрогать код, чтобы неиспользуемые фичи просто занимали объём флеша, а не оперативной памяти, то можно ограничиться парой-тройкой готовых наборов для каждой платформы, типа для обычной шторы, лампы или с упором на бегающие строки.

**Comment by vvip-68** [comment]
И да - последние "ардуиновские поделки IDE, которые и IDE то называть совестно" - это обычный VSCode со своими скинами и на(д)стройками. А Platformio - не на VSCode разве основано его IDE?

**Reply by hhrhhr** [reply]
Под VSCode есть официально поддерживаемое расширение, которое добавляет 5 кнопок на панель для компиляции-прошивки. Не проверял, но для ESP32 есть соединение с отладчиком, правда, точки останова или редактирование регистров всё равно нужно делать через консоль dbg.

Иначе говоря, VSCode используется только как редактор кода, с банальным отсутствием парсера определений (define). Если я в _a_def_hard.h_ делаю `#define USE_WEATHER 0`, например, то в других местах блоки #if-#else никак не реагируют на это.

Кое-как такое можно сделать в Codeblocks, но и там приходится все определения из кода тащить в конфиг проекта, чтобы редактор внятно подсвечивал (или нет) блоки, которые будут реально использованы.

Без VSCode можно сделать `pio home`, после чего утилита поднимает веб-сервер для [управления проектами](https://docs.platformio.org/en/latest/home/index.html#piohome). Технически, достаточно блокнота и этой штуки, разве что открывать файлы и переходить на нужную строку придётся ручками.

**Reply by vvip-68** [reply]
Займетесь адаптацией проекта - как установить и запустить его на pio, не меняя самого кода? Напишите инструкцию как этим пользоваться? Я добавлю необходимые файлы в проект.

Ну или сделаете может быть форк, чтобы все работало как вы хотите? Я в вики и на первую страничку моего репозитория помещу ссылку на ваш форк.

**Reply by hhrhhr** [reply]
Да занялся уже, себе исключительно через pio собираю, новый салют пытаюсь впихнуть. Форкну, оформлю новую страничку в вики и покажу.

Но встретился с бедой, если сделать `USE_E131 0`, то в Arduino IDE прошивка собирается как обычно, а при сборке через pio сыплет ошибками типа "необьявленная функция/структура e131_packet_t" и тп. Переключение на однопоточную компиляцию не спасает, почему-то pio пихает код из _e131.ino_ раньше, чем _a_def_hard.h_, либо не обращяет внимание на define.
Хотя вроде бы все использования функций из E1.31 обёрнуты дефайнами и не должны использоваться. 

Если сделать из _e131.ino_ нормальную пару cpp/h, с обявлениями, и подключить заголовки где надо, собирается без проблем. Но это уже не попадает под "не меняя самого кода" ;) Как победю, так для 8266 конфиг будет, благо что для ESP32 уже точно есть готовый.

**Reply by vvip-68** [reply]
Возможно потому, что pio "сам подтягивает нужные библиотеки"? Потому что именно эта библиотека - правленная. Не помню уже точно что - но в стандартной не хватало чего-то для корректной работы - пришлось дописывать. Возможно как раз этот тип дописался. А в штатной его просто нет.

**Reply by hhrhhr** [reply]
Так в том и дело, что ESP_Async_E1.31/ESPAsyncUDP не подтягиваются, то есть парсер заголовков понимает, что библиотека не используется, да и в списке lib_deps её нет. А компилятору почему-то передаются файлы и куски кода использующие E1.31, даже до линковщика дело не доходит, вот подай ему структуру e131_packet_t и всё тут.
При этом если не я использую MP3-плеер (аналогичная ситуация), то не одной функции оттуда не тащится. В общем, порядок включения исходных файлов можно как-то принудительно задать, осталось только понять как именно )

**Comment by vortigont** [comment]
вот здесь адаптированная под платформио ветка с актуальными либами если кому интересно
https://github.com/vortigont/LedPanelWiFi/tree/pio

**Reply by vvip-68** [reply]
Сразу отмечу, если будут какие-то вопросы по ЭТОМУ варианту проекта - вопросы задавайте держателю репозитория этой адаптированной версии. Во первых - код скетча там годовалой давности - с тех пор много чего было исправлено и дополнено. Во вторых - в dependencies библиотека EspAsyncWebServer там какая-то неизвестная мне, но я уже наступал на ее грабли - с этой версией библиотеки соединение по WebSocket постоянно рвется и в интерфейсе постоянные "Соединение потеряно" и невозможно ничего настоить/управлять...

**Reply by vortigont** [reply]
да я вроде здесь ничего не продаю. Год назад собирал ваш проект на платформио, так и осталось. Влил актуальные изменения. То что Вам сборка на платформио не нужна вы еще тогда сказали. Кому-то если пригодится настроеный `platfomio.ini` - пожалуйста.
А насчёт "неизвестной вам" версии асинк-сервера с граблями уж постыдились бы... эта [версия](https://github.com/mathieucarbou/ESPAsyncWebServer/graphs/contributors) на сегодняшний день единственная которая находится в активной поддержке, развивается и оптимизируется практически [ежедневно](https://github.com/mathieucarbou/ESPAsyncWebServer/graphs/commit-activity) и покрыта огромным количеством тестов, как синтетических так и в рамках проектов на её основе.
Ну вам виднее, опять же я ничего не продаю

**Reply by vvip-68** [reply]
Я просто, чтобы ко мне вопросов не было, если что пойдет не так. Я подсказать не смогу - сразу предупредил.
По версии библиотеки сервера - не всегда самое последнее самое хорошее и вообще подходящее для проекта.
Хотите экспериментировать - да ради Бога...  Почему я об этом сказал - с вашей же подачи брал эту библу в проект, как самую поддерживаемую и самую актуальную. Начали сыпаться сообщения что не могут подключиться к веб-страничке - постоянные отвалы. Собрал скомпилил у себя - действительно связь постоянно рвется. Два дня бился - безрезультатно. Даже проект с гитхаба снес, как неудачный и нерабочий.

Потом посмотрел, что версия сервера стоит не та, что у меня изначально была - видно когда-то экспериментировал, потом не удалил. Восстановил изначально использовавшуюся библиотеку - и о чуда - все обрывы прекратились. Потом пока шарился выяснял что не так, почему перестало работать - где-то мельком видел, кто как раз в этой библе какой-то постоянный принципиальный баг (или особенность), что она рвет соединение Web-сокета при каких-то условиях (уже не помню) которые у меня в проекте как раз постоянное явление. То ли пинг идет то ли какой-то параметр меняется для управления.

Поэтому все что сказано - из личного опыта. Просто предупредил.

**Reply by vortigont** [reply]
> с вашей же подачи брал эту библу в проект, как самую поддерживаемую и самую актуальную

ну я же вам её посоветовал попробовать не от того что она такая распрекрасная сама по себе а от того что у вас и до этого были проблемы со случайным полным отвалом вебморды с ваших же слов.

этот форк сейчас поддерживаем Метью и я, и за год мы её перетряхнули с головы до ног практически. В том числе и асинктцп на котором она работает. Все известные заявленные проблемы которые были изолированны до уровня библиотек - были решены. Дело не в том что новая - лучшая, дело в том какой код поддерживается, а какой нет.
Мы не занимаемся экспериментами по подбору "рабочей" версии, мы выявляем и решаем существующие проблемы.

**Reply by hhrhhr** [reply]
Конекретно по ESPAsyncWebServer - можно брать и пользовать, без подстройки чего-то типа `-D CONFIG_ASYNC_TCP_STACK_SIZE=4096`?

**Reply by vvip-68** [reply]
@vortigont - да не было проблем с подключением - вы что-то путаете. Вам тогда не понравилась процедура получения времени с ntp-серверов - вы ее мне переписали на использование системной в ядре. Спасибо. Работает. Проблемы с подключением начались как раз после смены библиотек.  Проблема, о которой, вы, возможно, говорите -выражалась в переполнении очереди  отправляемых в сокет сообщений. Вот с той, что сейчас стоит - более менее проблема устаканилась (плюс код тогда оптимизировал, чтобы уменьшить количество отправок? работу с памятью пересмотрели). А в поддерживаемой вами -библиотеке  проблема с переполнением -  в полный рост на том же коде. Я нашел вариант / комбинацию работающих библиотек и новую крутую с кучей неиспользуемых мной возможностей и отжирающей на них дополнительно память мне снова ставить и экспериментировать уж как-то не с руки. Извините. Впрочем я ж не отговариваю желающих поэкспериментировать - код открытый - кто хочет пусть дерзает. Только я на вопросы при использовании других вариантов отвечать не смогу и не буду.

**Reply by vvip-68** [reply]
Еще была проблема с мигающим первым светодиодом. А это никак с библиотекой сервера не связано. Это скорее проблемы с FastLED. Вы говорили, что у меня по-видимому проблемы с выделением / освобождением памяти. Но нет. Как оказалось - проблема известная. Об этом в Issues FastLED куча вопросов. И надежного решения как такового нет. Люди с осциллографом анализировали чколько данных уходит в линию. Библиотека 2.7.4 таких проблем не имеют, а в более поздних что-то переписали так, что теперь это постоянная проблема. Что-то у них конфликтует на одном ядре между отправкой данных в линию (критичную по временным интервалам) и работой стека WiFi.   Вон на версии 3.9.4 вообще к сети не может подключиться ESP8266

**Reply by vortigont** [reply]
> Конекретно по ESPAsyncWebServer - можно брать и пользовать, без подстройки чего-то типа `-D CONFIG_ASYNC_TCP_STACK_SIZE=4096`?

@hhrhhr можно, почему нет-то?
Надо просто понимать что и зачем подстраивать `CONFIG_ASYNC_TCP_STACK_SIZE` по-умолчанию исторически выставлен в 16к, это ОЧЕНЬ много для нормально спроектированного кода и попусту тратит память. Если не жалко или не уверены в своём коде то можно оставить как есть.

**Reply by hhrhhr** [reply]
> можно, почему нет-то? Надо просто понимать что и зачем подстраивать `CONFIG_ASYNC_TCP_STACK_SIZE` по-умолчанию исторически выставлен в 16к, это ОЧЕНЬ много

Когда я смотрю в лог при работающем веб-сервере гирлянды и вижу несколько сот байт свободной памяти, то тоже понимаю, что 16К это много. Вопрос скорее в разрезе использования с 8266, что можно, точнее **нужно**, убавить-выключить-ограничить.

**Reply by vvip-68** [reply]
Вот вероятно и ответ на вопрос, который у меня вчера спрашивали. Гирлянда на ESP8266. Свободной памяти остается 19-23К. Все работает нормально, пока не пытался подключиться к Web-серверу. Библиотека сервера как раз насколько я понял - стояла вот эта вот. При попытке подключиться к страничке контроллер тупо перезагружался. Вероятно, потому что веб-сервер пытался выделить память 16к по умолчанию из оставшихся 19 и оно все падало. Поменяли сервер из комплекта проекта - все наладилось. Соединение без проблем, всё работает... Кто ж знал, что там еще дополнительно что-то настраивать надо было, ключи вот эти где-то указывать... Товарищ 10 дней безуспешно пытался заставить работать проект. Пока не взял библиотеки из комплекта проекта.

**Reply by vortigont** [reply]
> Вопрос скорее в разрезе использования с 8266, что можно убавить-выключить-ограничить.

в разрезе использования 8266 можно убавить сам 8266 и перейти на семейство есп32. На сегодняшний день пытаться что-то выкроить на 8266 вместо того что бы взять нормальный контроллер нужно быть очень упоротым человеком.

**Reply by vortigont** [reply]
> Вероятно, потому что веб-сервер пытался выделить память 16к по умолчанию из оставшихся 19 и оно все падало

ерунду говорите какую-то. Под 8266 используется совсем другая библиотека асинктцп, там нет ни отдельного стека ни, соответсвенно, параметра задающего его размер.

**Reply by hhrhhr** [reply]
@vvip-68
> Вон на версии 3.9.4 вообще к сети не может подключиться ESP8266

Там же >3.9.3 вообще обьявлено как бета 4.0. Страшно такое на постоянку использовать, каждый день мощные движения в коде, что ломают вчерашние примеры из этого же самого кода.

**Reply by vvip-68** [reply]
@vortigont  ну это было просто предположение - попытка объяснить причину по которой с вашей библиотекой контроллер просто падает и перезагружается. Возможно не в этом причина - не спорю.

@hhrhhr на esp32 собирается и работает. Пока проблем не замечал. И эта версия даже в списке релизов есть...
А теперь смотрю еще и 3.9.5 и 3.9.6  как релизы выложили...

Вот пишут и горящий первый светодиод на ленте наконец, вроде, победили. Может теперь и соединяться с сетью на 8266 будет

Sticky first green LED on the chain has been fixed. It turned out to be aggressive RMT recycling. We've disabled this for now and filed a bug:

**Reply by hhrhhr** [reply]
> можно убавить сам 8266 и перейти на семейство есп32.

Иначе говоря, на либу можно не смотреть и не заморачиваться её "адаптацией". Тем более что на страничке проекта и так обьявлено о прекращении поддержки 8266 очень скоро.

> взять нормальный контроллер

Ну так такой контроллер и стоит как 2 ненормальных. Если уж развивать тему, то можно и малинку/апельсинку взять с линуксом, все эффекты на Lua/Python гонять и вообще не париться размером памяти $)

**Reply by vortigont** [reply]
> так такой контроллер и стоит как 2 ненормальных

esp32 lite, s2 mini и wemos 8266 стоят 200 рублей на али. Впрочем даже если бы есп32 стоила минимум 400 - решайте для себя сами что вам дороже, 200 рублей или куча гемора с морально устаревшим глюкавым контроллером. Особенно на фоне цены за ленты/матрицы.

**Reply by hhrhhr** [reply]
Глюкавость - это от неумения работать с памятью обычно, нехватка, сработка OOM, но проще контроллер назначить виновным.

Про 200 рублей может оно когда-то и было так. Сейчас немного иначе, и продаваны с единичными отзывами и доставка небесплатная
![Clipboard01](https://github.com/user-attachments/assets/dc3ca3f9-bd33-4d3e-a772-c1b4200e42b6)

Но если вопрос о настройке библиотеки на минимальное потребление ресурсов решается заменой контроллера, то такая библиотека явно не подходит к использованию.

**Comment by vortigont** [comment]
> А в поддерживаемой вами -библиотеке проблема с переполнением - в полный рост на том же коде

ну во-первых это не моя библиотека, я подключился к разработке форка Мэттью сравнительно недавно, уже сильно после наших прошлогодних обсуждений. Тогда этого форка вообще не было а был yobox-mod, который Мэттью и унаследовал. С тех пор там изменилось настолько много всего, что смысла обсуждать это в контесте ваших проблем годовалой давности нет никаких.

во-вторых та версия что у вас в репе это по сути та же сильно устаревшая исходная либа от Хриcто Гочкова и там во всю имеются все те же проблемы с очередью которые там были изначально

в-третьих что бы понять текущие состояние "моей либы" как вы говорите - поднят [вопрос](https://github.com/espressif/arduino-esp32/issues/10526#issuecomment-2441154510) о том что бы включить этот форк целиком в состав ардуино ядра

в-чертвертых если вам действительно интересен подробный анализ проблемы с "очередью" то может почитать [этот](https://github.com/mathieucarbou/ESPAsyncWebServer/discussions/165) тред. В кратце я буквально пару недель как добавил [механизмы](https://github.com/mathieucarbou/ESPAsyncWebServer/discussions/165#discussioncomment-11554123) управления очередью событий в асинктцп и контроля потоком в асинксервер только чтобы снизить влияние от коряво написанного юзеркода который дергается в коллбеках сервера и засирает очередь всяким мусором или тупо тормозит обработку событий из очереди что и приводит к её переполнению, неожиданно да... (на самом деле нет). Так что проблемы обычно оказываются с "другой" стороны либы. Мало кто может грамотно писать быстрый асинхронный код.

как-то так. Но если у вас есть проверенный набор либ который вас вполне устраивает то не стоит ничего менять, это базовый принцип "работает - не трогай!".

**Comment by vortigont** [comment]
> Но если вопрос о настройке библиотеки на минимальное потребление ресурсов решается заменой контроллера

@hhrhhr еще раз теперь вам - для 8266 используется совсем _другая_ библиотека асинктцп в которой отсутствует параметр `CONFIG_ASYNC_TCP_STACK_SIZE` ввиду отсутствия самого стека треда как такогого. Там нечего настраивать ни на минимальное ни на максимальное что бы то ни было.
Хотите себе гемора за 100 рублей - ради Бога, дерзайте и настраивате что вам только в голову взбредёт.

**Reply by hhrhhr** [reply]
[Вот тут](https://github.com/mathieucarbou/ESPAsyncWebServer?tab=readme-ov-file#important-recommendations) вообще-то ни слова про используемую платформу, да и название соответсвующее - Important recommendations.

Себе я хочу чтобы предлагаемая вами библиотека работала на 8266 ещё лучше, чем текущая. Иначе смысла в замене не видится.

**Reply by vortigont** [reply]
> [Вот тут](https://github.com/mathieucarbou/ESPAsyncWebServer?tab=readme-ov-file#important-recommendations) вообще-то ни слова про используемую платформу

ну там и про выбор ядра параметр есть что не имеет смысла для 8266. Асинк сервер состоит из двух частей - сам сервер, реализующий хттп и вебсокет и прослойки над TCP между приложением и ядром есп. У есп32 и 8266 ядра разные, соответсвенно прослойка тоже разная.

> Себе я хочу чтобы предлагаемая вами библиотека работала на 8266 ещё лучше, чем текущая. Иначе смысла в замене не видится.

замена смысла не имеет - если старая версия вас целиком и полностью устраивает, используйте её. Чуда не будет. На самом деле это заслуга Метью что он сохраняет совместимость с АПИ исходной либы и тащит 8266 до сих пор. Я уже давно предложил бросить это. 8266 это контроллер >10 летней давности и вообще является "первым блином" эспрессиф с огромным количеством нерешенных проблем. Эспрессиф даже намеренно выпустила чипы в семействе есп32 полностью совместимые с 8266 по размеру, распиновке и цене что бы стащить людей с этого "проходного" контроллера. У них есть обязательсва перед корпоративными заказчиками, которые уже наплодили под них устройства и софт и поэтому есть утвержденный срок жизни чипа. К любителям, ваяющим штучные поделки на макетках это не относится. Но всё равно находятся очень упоротые которые в 24м году продолжают сражаться с ветряными мельницами. Ваш выбор.
Никаких "улучшений" специально для 8266 не будет уже, ни в либе вебсервера, ни в асинтцпт ни в ядре ардуино. Всё что делается сейчас это в рамках разработки под есп32, если это что-то поправит на 8266 тоже - ну прекрасно, нет так нет. 4я версия сервера отбросит всё легаси вообще что бы сконцентрировать усилися (пока) двух разработчиков на актуальных задачах. Мэтью единственный кто за посление 3-4 года смог вложить достаточно усилий что бы комплексно поддерживать и сопровождать эту библиотеку. Все остальные форки это просто набор штучных затычек которые собирают по принципу "моё завелось и ладно".


> В общем, порядок включения исходных файлов можно как-то принудительно задать, осталось только понять как именно )

удачи вам на этом пути... я еще год назад предлагал Вадиму отрефакторить проект и организвать его в виде обычных заголовков и цпп файлов что для ардуино ИДЕ вполне нормально. Но он категорически отказался ссылаясь на то что ему так удобнее и его всё устраивает. Он крайне консервативный человек (не в обиду @vvip-68 ).
